File: 07/haskell/main.hs
-- Part 1 is just the mid point of the list. That makes sense from a math perspective

import Data.List
import Data.List.Split
import System.IO

getPositions :: String -> [Int]
getPositions i = read <$> splitOn "," i

getTotalCost :: [Int] -> Int -> Int
getTotalCost positions target = sum $ map (abs . (target -)) positions

getTotalCost' :: [Int] -> Int -> Int
getTotalCost' positions target = sum $ map (triangle . abs . (target -)) positions

triangle :: Int -> Int
triangle n = (n * (n + 1)) `div` 2

-- Brute force part 1
part1 :: [Int] -> Int
part1 positions = minimum $ map (getTotalCost positions) [minimum positions .. maximum positions]

-- Part 1 using median
part1' :: [Int] -> Int
part1' positions = getTotalCost positions (sort positions !! (length positions `div` 2))

-- Brute force part 2
part2 :: [Int] -> Int
part2 positions = minimum $ map (getTotalCost' positions) [minimum positions .. maximum positions]

-- Part 2 using mean (I don't really know why this is... would need to do some reading)
part2' :: [Int] -> Int
part2' positions = getTotalCost' positions $ sum positions `div` length positions

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print (part1' $ getPositions contents)
  print (part2' $ getPositions contents)
  hClose handle

File: 07/python/main.py
def solution(lines):
    levels = [int(x) for x in lines[0].split(',')]
    min_cost = None
    for i in range(0, max(levels)):
        total_cost = sum([(sum(z for z in range(1,abs(y - i)+1))) for y in levels])
        if min_cost is None or total_cost < min_cost:
            min_cost = total_cost
            print(i, total_cost)


with open('input.txt') as f:
    lines = f.readlines()
    print(solution(lines))